File: server-rust/src/backtest/broker.rs
//! 经纪商模拟器 — Decimal 精度
//!
//! 内置 A 股规则：T+1、涨跌停、100 股整数倍、印花税、过户费

use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use std::collections::HashMap;

use super::types::{Bar, Order, OrderType, Trade, AccountSnapshot};

/// 涨跌幅限制百分比
fn limit_pct(symbol: &str, is_st: bool) -> Decimal {
    if is_st {
        return dec!(0.05); // ST 股 5%
    }
    // 北交所 30%
    if symbol.starts_with('8') || symbol.starts_with('4') {
        return dec!(0.30);
    }
    // 科创板 688 / 创业板 30/301 → 20%
    if symbol.starts_with("688") || symbol.starts_with("30") {
        return dec!(0.20);
    }
    // 主板默认 10%
    dec!(0.10)
}

pub struct Broker {
    initial_cash: Decimal,
    cash: Decimal,
    commission_rate: Decimal,
    slippage: Decimal,
    stamp_duty_rate: Decimal,
    transfer_fee_rate: Decimal,
    enable_t1: bool,
    enable_limit: bool,

    positions: HashMap<String, Decimal>,       // symbol -> 持仓股数
    cost_basis: HashMap<String, Decimal>,       // symbol -> 成本价
    today_bought: HashMap<String, Decimal>,     // symbol -> 当日净买入量
    last_trade_date: Option<chrono::NaiveDate>,
    pending_orders: Vec<Order>,

    pub trades: Vec<Trade>,
    pub equity_curve: Vec<AccountSnapshot>,
    last_prices: HashMap<String, Decimal>,
}

impl Broker {
    pub fn new(
        cash: Decimal,
        commission_rate: Decimal,
        slippage: Decimal,
        stamp_duty_rate: Decimal,
        transfer_fee_rate: Decimal,
    ) -> Self {
        let s = Self {
            initial_cash: cash,
            cash,
            commission_rate,
            slippage,
            stamp_duty_rate,
            transfer_fee_rate,
            enable_t1: true,
            enable_limit: true,
            positions: HashMap::new(),
            cost_basis: HashMap::new(),
            today_bought: HashMap::new(),
            last_trade_date: None,
            pending_orders: Vec::new(),
            trades: Vec::new(),
            equity_curve: Vec::new(),
            last_prices: HashMap::new(),
        };
        println!("Broker::new positions_len={}", s.positions.len());
        s
    }

    pub fn with_options(mut self, enable_t1: bool, enable_limit: bool) -> Self {
        self.enable_t1 = enable_t1;
        self.enable_limit = enable_limit;
        self
    }

    /// 总资产 = 现金 + 持仓市值
    pub fn total_value(&self) -> Decimal {
        let pos_val: Decimal = self
            .positions
            .iter()
            .map(|(sym, qty)| qty * self.last_prices.get(sym).copied().unwrap_or(dec!(0)))
            .sum();
        self.cash + pos_val
    }

    pub fn position_value(&self) -> Decimal {
        self.positions
            .iter()
            .map(|(sym, qty)| qty * self.last_prices.get(sym).copied().unwrap_or(dec!(0)))
            .sum()
    }

    /// 获取某标的持仓股数
    pub fn position_qty(&self, symbol: &str) -> Decimal {
        self.positions.get(symbol).copied().unwrap_or(dec!(0))
    }

    /// 是否持有任何仓位
    pub fn has_positions(&self) -> bool {
        let result = !self.positions.is_empty();
        println!("has_positions: len={}, result={}", self.positions.len(), result);
        result
    }

    /// 提交订单到待执行队列
    pub fn submit_order(&mut self, order: Order) {
        self.pending_orders.push(order);
    }

    /// 执行所有待处理订单
    pub fn execute_pending_orders(&mut self, bars: &[Bar]) {
        let orders: Vec<Order> = self.pending_orders.drain(..).collect();
        self.execute(&orders, bars);
    }

    /// 执行订单
    pub fn execute(&mut self, orders: &[Order], bars: &[Bar]) {
        println!("execute: orders={}, bars={}", orders.len(), bars.len());
        // 交易日切换检测
        if self.enable_t1 && !bars.is_empty() {
            let current_date = bars[0].timestamp.date();
            if self.last_trade_date != Some(current_date) {
                self.today_bought.clear();
                self.last_trade_date = Some(current_date);
            }
        }

        // 建立 symbol -> bar 映射
        let bar_map: HashMap<&str, &Bar> = bars.iter().map(|b| (b.symbol.as_str(), b)).collect();
        println!("bar_map keys: {:?}", bar_map.keys().collect::<Vec<_>>());

        for order in orders {
            println!("Processing order: {} {:?} amount={}", order.symbol, order.order_type, order.amount);
            let Some(bar) = bar_map.get(order.symbol.as_str()) else {
                println!("  Bar not found for {}", order.symbol);
                continue;
            };

            let close = bar.close;
            let pre_close = if bar.pre_close > dec!(0) {
                bar.pre_close
            } else {
                close
            };

            self.last_prices.insert(order.symbol.clone(), close);

            // 涨跌停检查
            if self.enable_limit && pre_close > dec!(0) {
                let pct = limit_pct(&order.symbol, bar.is_st);
                let limit_up = pre_close * (Decimal::ONE + pct);
                let limit_down = pre_close * (Decimal::ONE - pct);

                match order.order_type {
                    OrderType::Buy if close >= limit_up * dec!(0.999) => continue,
                    OrderType::Sell if close <= limit_down * dec!(1.001) => continue,
                    _ => {}
                }
            }

            // 滑点
            let fill_price = match order.order_type {
                OrderType::Buy => close * (Decimal::ONE + self.slippage),
                OrderType::Sell => close * (Decimal::ONE - self.slippage),
            };

            let amount = Decimal::from(order.amount);
            let cost = fill_price * amount;

            let commission = (cost * self.commission_rate)
                .round_dp_with_strategy(2, rust_decimal::RoundingStrategy::MidpointNearestEven);
            let transfer_fee = (cost * self.transfer_fee_rate)
                .round_dp_with_strategy(2, rust_decimal::RoundingStrategy::MidpointNearestEven);

            match order.order_type {
                OrderType::Buy => {
                    let total_cost = cost + commission + transfer_fee;
                    println!("  Buy: cash={}, total_cost={}", self.cash, total_cost);
                    if self.cash < total_cost {
                        println!("  Buy: cash insufficient");
                        continue;
                    }
                    self.cash -= total_cost;
                    println!("  Buy: cash after={}", self.cash);

                    let old_qty = self.positions.get(&order.symbol).copied().unwrap_or(dec!(0));
                    let old_cost = self.cost_basis.get(&order.symbol).copied().unwrap_or(dec!(0)) * old_qty;
                    let new_qty = old_qty + amount;
                    if new_qty > dec!(0) {
                        self.cost_basis.insert(order.symbol.clone(), (old_cost + cost) / new_qty);
                    } else {
                        self.cost_basis.insert(order.symbol.clone(), dec!(0));
                    }
                    self.positions.insert(order.symbol.clone(), new_qty);
                    println!("  Buy: positions inserted, trades before push={}", self.trades.len());

                    *self.today_bought.entry(order.symbol.clone()).or_insert(dec!(0)) += amount;

                    self.trades.push(Trade {
                        symbol: order.symbol.clone(),
                        amount: order.amount,
                        price: fill_price,
                        order_type: OrderType::Buy,
                        timestamp: order.timestamp,
                        commission,
                        stamp_duty: dec!(0),
                        transfer_fee,
                        pnl: dec!(0),
                    });
                    println!("  Buy: trades after push={}", self.trades.len());
                }
                OrderType::Sell => {
                    let current_qty = self.positions.get(&order.symbol).copied().unwrap_or(dec!(0));
                    if current_qty < amount {
                        continue;
                    }

                    // T+1: 当日买入部分不可卖出
                    if self.enable_t1 {
                        let today_bought = self.today_bought.get(&order.symbol).copied().unwrap_or(dec!(0));
                        let max_sellable = current_qty - today_bought;
                        if amount > max_sellable {
                            continue;
                        }
                    }

                    let stamp_duty = (cost * self.stamp_duty_rate)
                        .round_dp_with_strategy(2, rust_decimal::RoundingStrategy::MidpointNearestEven);
                    self.cash += cost - commission - stamp_duty - transfer_fee;

                    let avg_cost = self.cost_basis.get(&order.symbol).copied().unwrap_or(dec!(0));
                    let pnl = (fill_price - avg_cost) * amount - commission - stamp_duty - transfer_fee;

                    let new_qty = current_qty - amount;
                    if new_qty <= dec!(0) {
                        self.positions.remove(&order.symbol);
                        self.cost_basis.remove(&order.symbol);
                    } else {
                        self.positions.insert(order.symbol.clone(), new_qty);
                    }

                    self.trades.push(Trade {
                        symbol: order.symbol.clone(),
                        amount: order.amount,
                        price: fill_price,
                        order_type: OrderType::Sell,
                        timestamp: order.timestamp,
                        commission,
                        stamp_duty,
                        transfer_fee,
                        pnl,
                    });
                }
            }
        }
    }

    /// 记录净值快照
    pub fn record_snapshot(&mut self, timestamp: chrono::NaiveDateTime) {
        self.equity_curve.push(AccountSnapshot {
            timestamp,
            total_value: self.total_value(),
            cash: self.cash,
            position_value: self.position_value(),
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;

    fn make_broker() -> Broker {
        Broker::new(
            dec!(1_000_000),
            dec!(0.0003),
            dec!(0.001),
            dec!(0.0005),
            dec!(0.00001),
        )
    }

    fn bar(symbol: &str, close: Decimal, pre_close: Decimal) -> Bar {
        let ts = NaiveDate::from_ymd_opt(2023, 1, 5)
            .unwrap()
            .and_hms_opt(0, 0, 0)
            .unwrap();
        Bar {
            symbol: symbol.to_string(),
            timestamp: ts,
            open: close,
            high: close,
            low: close,
            close,
            volume: 1000000,
            pre_close,
            is_st: false,
        }
    }

    #[test]
    fn test_buy_and_cash_reduction() {
        let mut b = make_broker();
        let order = Order {
            symbol: "000001.SZ".to_string(),
            amount: 1000,
            order_type: OrderType::Buy,
            timestamp: b.equity_curve.first().map(|s| s.timestamp).unwrap_or(
                NaiveDate::from_ymd_opt(2023, 1, 5).unwrap().and_hms_opt(0,0,0).unwrap()
            ),
        };
        b.execute(&[order], &[bar("000001.SZ", dec!(100), dec!(99))]);
        assert_eq!(b.positions.get("000001.SZ").copied().unwrap_or(dec!(0)), dec!(1000));
        assert!(b.cash < dec!(1_000_000));
        assert_eq!(b.trades.len(), 1);
    }

    #[test]
    fn test_t1_blocks_same_day_sell() {
        let mut b = make_broker();
        let ts = NaiveDate::from_ymd_opt(2023, 1, 5).unwrap().and_hms_opt(0,0,0).unwrap();
        let buy = Order { symbol: "000001.SZ".to_string(), amount: 1000, order_type: OrderType::Buy, timestamp: ts };
        let sell = Order { symbol: "000001.SZ".to_string(), amount: 1000, order_type: OrderType::Sell, timestamp: ts };
        let the_bar = bar("000001.SZ", dec!(100), dec!(99));
        b.execute(&[buy], &[the_bar.clone()]);
        b.execute(&[sell], &[the_bar]);
        // 买入成交，卖出被 T+1 拦截
        let sell_trades: Vec<_> = b.trades.iter().filter(|t| matches!(t.order_type, OrderType::Sell)).collect();
        assert!(sell_trades.is_empty());
    }

    #[test]
    fn test_limit_up_blocks_buy() {
        let mut b = make_broker();
        let ts = NaiveDate::from_ymd_opt(2023, 1, 5).unwrap().and_hms_opt(0,0,0).unwrap();
        // pre_close=100, close=110 (涨停 10%)
        let up_bar = Bar { close: dec!(110), pre_close: dec!(100), ..bar("000001.SZ", dec!(110), dec!(100)) };
        let order = Order { symbol: "000001.SZ".to_string(), amount: 1000, order_type: OrderType::Buy, timestamp: ts };
        b.execute(&[order], &[up_bar]);
        assert!(b.trades.is_empty());
    }

    #[test]
    fn test_lot_size_must_be_100_multiple() {
        // Rust 中 amount 类型是 u64，但策略代码可能传入非 100 倍数
        // 这个测试验证引擎层面是否拒绝
        // 实际中 Context.buy 应该先把 amount 向下取整到 100 的倍数
        // 这里先测试 broker 接收 100 倍数能正常成交
        let mut b = make_broker();
        let ts = NaiveDate::from_ymd_opt(2023, 1, 5).unwrap().and_hms_opt(0,0,0).unwrap();
        let order = Order { symbol: "000001.SZ".to_string(), amount: 500, order_type: OrderType::Buy, timestamp: ts };
        b.execute(&[order], &[bar("000001.SZ", dec!(100), dec!(99))]);
        assert_eq!(b.positions.get("000001.SZ").copied().unwrap_or(dec!(0)), dec!(500));
    }
}

File: 项目废弃的python/server/backtest/engine/broker.py
"""
经纪商模拟器：模拟A股交易环境，支持滑点、佣金、印花税、持仓管理。
兼容 Python 3.11.8
核心资金计算使用 Decimal 保证精度。
"""
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Dict
from decimal import Decimal, ROUND_HALF_UP
from enum import Enum
import pandas as pd


class OrderType(Enum):
    BUY = "buy"
    SELL = "sell"


@dataclass
class Order:
    symbol: str
    amount: float
    price: float
    order_type: OrderType
    timestamp: datetime
    filled: bool = False
    fill_price: float = 0.0


@dataclass
class Trade:
    symbol: str
    amount: float
    price: float
    order_type: OrderType
    timestamp: datetime
    commission: float = 0.0
    pnl: float = 0.0


class Broker:
    def __init__(
        self,
        cash: float,
        commission: float = 0.0003,
        slippage: float = 0.001,
        stamp_duty: float = 0.0005,
        transfer_fee: float = 0.00001,
        enable_t1: bool = True,
        enable_limit: bool = True,
    ):
        self._initial_cash = Decimal(str(cash))
        self._cash = Decimal(str(cash))
        self._commission_rate = Decimal(str(commission))
        self._slippage = Decimal(str(slippage))
        self._stamp_duty_rate = Decimal(str(stamp_duty))      # 印花税：仅卖出收取
        self._transfer_fee_rate = Decimal(str(transfer_fee))  # 过户费：双向收取
        self.enable_t1 = enable_t1
        self.enable_limit = enable_limit
        self._positions: Dict[str, Decimal] = {}
        self.pending_orders: List[Order] = []
        self.trades: List[Trade] = []
        self._last_price: Dict[str, Decimal] = {}
        self._cost_basis: Dict[str, Decimal] = {}
        self._today_buy_amounts: Dict[str, Decimal] = {}  # symbol -> 当日净买入量（T+1用）
        self._last_trade_date = None

    @property
    def initial_cash(self) -> float:
        return float(self._initial_cash)

    @property
    def cash(self) -> float:
        return float(self._cash)

    @property
    def total_value(self) -> float:
        """总资产 = 现金 + 所有持仓按最新价市值"""
        position_value = sum(
            qty * self._last_price.get(sym, Decimal("0"))
            for sym, qty in self._positions.items()
        )
        return float(self._cash + position_value)

    @property
    def position_value(self) -> float:
        return float(sum(
            qty * self._last_price.get(sym, Decimal("0"))
            for sym, qty in self._positions.items()
        ))

    @property
    def positions(self) -> Dict[str, float]:
        return {sym: float(qty) for sym, qty in self._positions.items()}

    @staticmethod
    def _get_limit_pct(bar: pd.DataFrame, symbol: str) -> Decimal:
        """
        获取单只股票的涨跌幅限制。
        优先级：
        1. bar 数据中显式指定的 limit_pct 列
        2. ST 股（含 *ST）→ 5%
        3. 基于代码前缀推断板块
        """
        if "limit_pct" in bar.columns:
            raw = bar["limit_pct"].values[0]
            if raw is not None and raw == raw:  # 排除 NaN
                return Decimal(str(raw))
        # ST 股（含 *ST）：无论板块统一 5%
        if "is_st" in bar.columns:
            if bool(bar["is_st"].values[0]):
                return Decimal("0.05")
        # 北交所: 8xxxxx / 4xxxxx → 30%
        if symbol.startswith("8") or symbol.startswith("4") or symbol.startswith("83") or symbol.startswith("43") or symbol.startswith("87"):
            return Decimal("0.3")
        # 科创板: 688xxx → 20%
        if symbol.startswith("688"):
            return Decimal("0.2")
        # 创业板: 300xxx / 301xxx → 20%
        if symbol.startswith("30"):
            return Decimal("0.2")
        # 科创板/创业板注册制新股前5日无涨跌幅限制，但回测中简化处理
        # 主板默认 10%
        return Decimal("0.1")

    def submit_order(self, order: Order) -> None:
        """提交订单到待执行队列"""
        self.pending_orders.append(order)

    def execute_orders(self, bar_group: pd.DataFrame) -> None:
        """根据当前Bar数据执行所有待处理订单（含A股T+1、涨跌停限制）"""
        # 统一处理交易日切换：新交易日开始时清空当日买入量记录
        if self.enable_t1 and not bar_group.empty:
            current_date = pd.to_datetime(bar_group["datetime"].iloc[0]).date()
            if self._last_trade_date != current_date:
                self._today_buy_amounts.clear()
                self._last_trade_date = current_date

        for order in list(self.pending_orders):
            bar = bar_group[bar_group['symbol'] == order.symbol]
            if bar.empty:
                continue

            close = Decimal(str(bar['close'].values[0]))
            pre_close = Decimal(str(bar['pre_close'].values[0])) if 'pre_close' in bar.columns else close
            self._last_price[order.symbol] = close

            # A股涨跌停限制
            if self.enable_limit and pre_close > 0:
                limit_pct = self._get_limit_pct(bar, order.symbol)
                limit_up = pre_close * (Decimal("1") + limit_pct)
                limit_down = pre_close * (Decimal("1") - limit_pct)

                if order.order_type == OrderType.BUY and close >= limit_up * Decimal("0.999"):
                    # 涨停时无法买入（留出微小容差）
                    continue
                if order.order_type == OrderType.SELL and close <= limit_down * Decimal("1.001"):
                    # 跌停时无法卖出
                    continue

            # 模拟滑点
            if order.order_type == OrderType.BUY:
                fill_price = close * (Decimal("1") + self._slippage)
            else:
                fill_price = close * (Decimal("1") - self._slippage)

            amount = Decimal(str(order.amount))
            cost = fill_price * amount
            commission = (cost * self._commission_rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            transfer_fee = (cost * self._transfer_fee_rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

            # 卖出前记录成本价（在更新仓位前，因为全仓卖出会删除 _cost_basis）
            avg_cost_before = self._cost_basis.get(order.symbol)

            if order.order_type == OrderType.BUY:
                total_cost = cost + commission + transfer_fee
                if self._cash < total_cost:
                    continue
                self._cash -= total_cost
                old_qty = self._positions.get(order.symbol, Decimal("0"))
                old_cost = self._cost_basis.get(order.symbol, Decimal("0")) * old_qty
                new_qty = old_qty + amount
                if new_qty > 0:
                    self._cost_basis[order.symbol] = (old_cost + cost) / new_qty
                else:
                    self._cost_basis[order.symbol] = Decimal("0")
                self._positions[order.symbol] = new_qty
                self._today_buy_amounts[order.symbol] = self._today_buy_amounts.get(order.symbol, Decimal("0")) + amount
            else:
                current_qty = self._positions.get(order.symbol, Decimal("0"))
                if current_qty < amount:
                    continue

                # T+1 限制：当日买入的部分不可卖出，旧持仓可以卖出
                if self.enable_t1:
                    today_bought = self._today_buy_amounts.get(order.symbol, Decimal("0"))
                    max_sellable = current_qty - today_bought
                    if amount > max_sellable:
                        continue

                stamp_duty = (cost * self._stamp_duty_rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
                self._cash += cost - commission - stamp_duty - transfer_fee
                new_qty = current_qty - amount
                if new_qty <= 0:
                    del self._positions[order.symbol]
                    del self._cost_basis[order.symbol]
                else:
                    self._positions[order.symbol] = new_qty

            # 计算单笔PNL（扣除全部交易成本）
            pnl = Decimal("0")
            total_fees = commission + transfer_fee
            if order.order_type == OrderType.SELL:
                stamp_duty = (cost * self._stamp_duty_rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
                total_fees += stamp_duty
                if avg_cost_before is not None:
                    pnl = (fill_price - avg_cost_before) * amount - total_fees

            order.filled = True
            order.fill_price = float(fill_price)
            self.trades.append(Trade(
                symbol=order.symbol,
                amount=order.amount,
                price=float(fill_price),
                order_type=order.order_type,
                timestamp=order.timestamp,
                commission=float(total_fees),
                pnl=float(pnl),
            ))

        # 清除已成交订单
        self.pending_orders = [o for o in self.pending_orders if not o.filled]
